enterprise/coderd/prebuilds/reconcile.go

```diff
@@ -0,0 +1,511 @@
+		}
+
+		now := c.clock.Now()
+
+		minimumWorkspace, err := db.InsertWorkspace(ctx, database.InsertWorkspaceParams{
+			ID:                prebuildID,
+			CreatedAt:         now,
+			UpdatedAt:         now,
+			OwnerID:           prebuilds.SystemUserID,
+			OrganizationID:    template.OrganizationID,
+			TemplateID:        template.ID,
+			Name:              name,
+			LastUsedAt:        c.clock.Now(),
+			AutomaticUpdates:  database.AutomaticUpdatesNever,
+			AutostartSchedule: sql.NullString{},
+			Ttl:               sql.NullInt64{},
+			NextStartAt:       sql.NullTime{},
+		})
+		if err != nil {
+			return xerrors.Errorf("insert workspace: %w", err)
+		}
+
+		// We have to refetch the workspace for the joined in fields.
+		workspace, err := db.GetWorkspaceByID(ctx, minimumWorkspace.ID)
+		if err != nil {
+			return xerrors.Errorf("get workspace by ID: %w", err)
+		}
+
+		c.logger.Info(ctx, "attempting to create prebuild", slog.F("name", name),
+			slog.F("workspace_id", prebuildID.String()), slog.F("preset_id", presetID.String()))
+
+		return c.provision(ctx, db, prebuildID, template, presetID, database.WorkspaceTransitionStart, workspace)
+	}, &database.TxOptions{
+		Isolation: sql.LevelRepeatableRead,
+		ReadOnly:  false,
+	})
+}
+
+func (c *StoreReconciler) deletePrebuild(ctx context.Context, prebuildID uuid.UUID, templateID uuid.UUID, presetID uuid.UUID) error {
+	return c.store.InTx(func(db database.Store) error {
+		workspace, err := db.GetWorkspaceByID(ctx, prebuildID)
+		if err != nil {
+			return xerrors.Errorf("get workspace by ID: %w", err)
+		}
+
+		template, err := db.GetTemplateByID(ctx, templateID)
+		if err != nil {
+			return xerrors.Errorf("failed to get template: %w", err)
+		}
+
+		c.logger.Info(ctx, "attempting to delete prebuild",
+			slog.F("workspace_id", prebuildID.String()), slog.F("preset_id", presetID.String()))
+
+		return c.provision(ctx, db, prebuildID, template, presetID, database.WorkspaceTransitionDelete, workspace)
+	}, &database.TxOptions{
+		Isolation: sql.LevelRepeatableRead,
+		ReadOnly:  false,
+	})
+}
+
+func (c *StoreReconciler) provision(
+	ctx context.Context,
+	db database.Store,
+	prebuildID uuid.UUID,
+	template database.Template,
+	presetID uuid.UUID,
+	transition database.WorkspaceTransition,
+	workspace database.Workspace,
+) error {
+	tvp, err := db.GetPresetParametersByTemplateVersionID(ctx, template.ActiveVersionID)
+	if err != nil {
+		return xerrors.Errorf("fetch preset details: %w", err)
+	}
+
+	var params []codersdk.WorkspaceBuildParameter
+	for _, param := range tvp {
+		// TODO: don't fetch in the first place.
+		if param.TemplateVersionPresetID != presetID {
+			continue
+		}
+
+		params = append(params, codersdk.WorkspaceBuildParameter{
+			Name:  param.Name,
+			Value: param.Value,
+		})
+	}
+
+	builder := wsbuilder.New(workspace, transition).
+		Reason(database.BuildReasonInitiator).
+		Initiator(prebuilds.SystemUserID).
+		ActiveVersion().
+		VersionID(template.ActiveVersionID).
```

Comment:
ActiveVersion() and VersionID are mutually exclusive.

Reply:
I decided to keep `VersionID(template.ActiveVersionID).` to prevent additional DB query.